include/matx/transforms/matmul/matmul_cusparse.h

```diff
@@ -0,0 +1,327 @@
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+// OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+/////////////////////////////////////////////////////////////////////////////////
+
+#pragma once
+
+#include <cusparse.h>
+
+#include <cstdio>
+#include <numeric>
+
+#include "matx/core/cache.h"
+#include "matx/core/sparse_tensor.h"
+#include "matx/core/tensor.h"
+#include "matx/transforms/matmul/matmul_common.h"
+
+namespace matx {
+
+namespace detail {
+
+// Translate MatXType for indices to cuSPARSE index type.
+template <typename T>
+constexpr cusparseIndexType_t MatXTypeToCuSparseIndexType() {
+  if constexpr (std::is_same_v<T, uint16_t>) {
+    return CUSPARSE_INDEX_16U;
+  }
+  if constexpr (std::is_same_v<T, int32_t>) {
+    return CUSPARSE_INDEX_32I;
+  }
+  if constexpr (std::is_same_v<T, int64_t>) {
+    return CUSPARSE_INDEX_64I;
+  }
+  if constexpr (std::is_same_v<T, index_t>) {
+    return CUSPARSE_INDEX_64I;
+  }
+}
+
+/**
+ * Parameters needed to execute a cuSPARSE GEMM.
+ */
+struct MatMulCUSPARSEParams_t {
+  MatXDataType_t dtype;
+  MatXDataType_t ptype;
+  MatXDataType_t ctype;
+  int rank;
+  cudaStream_t stream;
+  float alpha;
+  float beta;
+  index_t nse;
+  index_t m;
+  index_t n;
+  index_t k;
+  cusparseOperation_t opA;
+  cusparseOperation_t opB;
+  // Matrix handles in cuSPARSE are data specific (unlike e.g. cuBLAS
+  // where the same plan can be shared between different data buffers).
+  void *ptrA0;
+  void *ptrA1;
+  void *ptrA2;
+  void *ptrA3;
+  void *ptrA4;
+  void *ptrB;
+  void *ptrC;
+};
+
+template <typename TensorTypeC, typename TensorTypeA, typename TensorTypeB>
+class MatMulCUSPARSEHandle_t {
+public:
+  using TA = typename TensorTypeA::value_type;
+  using TB = typename TensorTypeB::value_type;
+  using TC = typename TensorTypeC::value_type;
+
+  static constexpr int RANKA = TensorTypeC::Rank();
+  static constexpr int RANKB = TensorTypeC::Rank();
+  static constexpr int RANKC = TensorTypeC::Rank();
+
+  /**
+   * Construct a sparse GEMM handle
+   *   SpMV
+   *   SpMM        <- for now
+   *   SpGEMM
+   *
+   */
+  MatMulCUSPARSEHandle_t(TensorTypeC &c, const TensorTypeA &a,
+                         const TensorTypeB &b, cudaStream_t stream, float alpha,
+                         float beta) {
+    MATX_NVTX_START("", matx::MATX_NVTX_LOG_INTERNAL)
+
+    static_assert(RANKA == 2);
+    static_assert(RANKB == 2);
+    static_assert(RANKC == 2);
+
+    MATX_ASSERT(a.Size(RANKA - 1) == b.Size(RANKB - 2), matxInvalidSize);
+    MATX_ASSERT(c.Size(RANKC - 1) == b.Size(RANKB - 1), matxInvalidSize);
+    MATX_ASSERT(c.Size(RANKC - 2) == a.Size(RANKA - 2), matxInvalidSize);
+
+    params_ = GetGemmParams(c, a, b, stream, alpha, beta);
+
+    cusparseStatus_t ret = cusparseCreate(&handle_);
```

Comment:
All the places `ret` is only used in an assert needs `[[maybe_unused]]` before it since it's disabled in release mode right now